src/basedtyping/transformer.py

```diff
@@ -0,0 +1,231 @@
+        if isinstance(node_type, Enum):
+            assert isinstance(node, (ast.Name, ast.Attribute))
+            return self._literal(node)
+        return node
+
+    def visit_Name(self, node: ast.Name) -> ast.AST:
+        name_type = self.eval_type(node)
+        if isinstance(name_type, Enum):
+            return self._literal(node)
+        return node
+
+    def visit_Constant(self, node: ast.Constant) -> ast.AST:
+        value = cast(object, node.value)
+        if isinstance(value, int) or (self.string_literals and isinstance(value, str)):
+            return self._literal(node)
+        return node
+
+    def visit_Tuple(self, node: ast.Tuple) -> ast.AST:
+        node = self.generic_visit(node)
+        if not self._implicit_tuple:
+            return self.subscript(self._typing("Tuple"), cast(ast.expr, node))
+        return node
+
+    def visit_Compare(self, node: ast.Compare) -> ast.AST:
+        if len(node.ops) == 1 and isinstance(node.ops[0], ast.Is):
+            result = self.subscript(
+                self._typing("TypeIs"), cast(ast.expr, self.generic_visit(node.comparators[0]))
+            )
+            return self.generic_visit(result)
+        return self.generic_visit(node)
+
+    def visit_IfExp(self, node: ast.IfExp) -> ast.AST:
+        if (
+            isinstance(node.body, ast.Compare)
+            and len(node.body.comparators) == 1
+            and isinstance(node.body.ops[0], ast.Is)
+        ):
+            node.body = self.subscript(
+                self._typing("TypeGuard"),
+                cast(ast.expr, self.generic_visit(node.body.comparators[0])),
+            )
+        return self.generic_visit(node)
+
+    def visit_FunctionType(self, node: ast.FunctionType) -> ast.AST:
+        node = self.generic_visit(node)
+        assert isinstance(node, ast.FunctionType)
+        return self.subscript(
+            self._typing("Callable"),
+            ast.Tuple([ast.List(node.argtypes, ctx=ast.Load()), node.returns], ctx=ast.Load()),
+        )
+
+    def visit_BinOp(self, node: ast.BinOp) -> ast.AST:
+        node = self.generic_visit(node)
+        if isinstance(node, ast.BinOp) and isinstance(node.op, ast.BitAnd):
+            node = self.subscript(
+                self._basedtyping("Intersection"),
+                ast.Tuple([node.left, node.right], ctx=ast.Load()),
+            )
+        return node
+
+
+def _eval_direct(
+    value: typing.ForwardRef,
+    globalns: dict[str, object] | None = None,
+    localns: dict[str, object] | None = None,
+) -> object:
+    return eval_type_based(value, globalns, localns, string_literals=False)
+
+
+def eval_type_based(
+    value: typing.ForwardRef,
+    globalns: dict[str, object] | None = None,
+    localns: dict[str, object] | None = None,
+    *,
+    string_literals: bool,
+) -> object:
+    """Like `typing._eval_type`, but supports based typing features.
+    Specifically, this transforms `1 | 2` into `typing.Union[Literal[1], Literal[2]]`
```

Comment:
why not
```suggestion
    Specifically, this transforms `1 | 2` into `Literal[1, 2]`
```

Reply:
it transforms it in to the old version, to support old pythons

Reply:
old versions of python dont support `Literal[1, 2]`?

Reply:
oops, it doesn't convert new to old unions...

i mean, it's much simpler to convert `1` to `Literal[1]`, instead of performing semantic analysis on the entire expressions and discovering where and how to perform any kind of simplifications

Reply:
`get_args` doesn't work with `Literal` unions (directly contradicting the typing spec BTW), but nested literals are automatically unpacked, so those do work correctly.

So you could also do something like
```suggestion
    Specifically, this transforms `1 | 2` into `Literal[Literal[1], Literal[2]]`
```

to get the runtime equivalent of `Literal[1, 2]`